src/optimizer/levenberg_marquardt_optimizer.rs

```diff
@@ -0,0 +1,175 @@
+use log::trace;
+use std::ops::Mul;
+use std::{collections::HashMap, time::Instant};
+
+use faer_ext::IntoNalgebra;
+
+use crate::common::OptimizerOptions;
+use crate::linear;
+use crate::optimizer;
+use crate::sparse::LinearSolverType;
+use crate::sparse::SparseLinearSolver;
+
+#[derive(Debug)]
+pub struct LevenbergMarquardtOptimizer {}
+impl LevenbergMarquardtOptimizer {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+impl Default for LevenbergMarquardtOptimizer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl optimizer::Optimizer for LevenbergMarquardtOptimizer {
+    fn optimize(
+        &self,
+        problem: &crate::problem::Problem,
+        initial_values: &std::collections::HashMap<String, nalgebra::DVector<f64>>,
+        optimizer_option: Option<OptimizerOptions>,
+    ) -> Option<HashMap<String, nalgebra::DVector<f64>>> {
+        let mut params = initial_values.clone();
+        let opt_option = optimizer_option.unwrap_or_default();
+        let mut linear_solver: Box<dyn SparseLinearSolver> = match opt_option.linear_solver_type {
+            LinearSolverType::SparseCholesky => Box::new(linear::SparseCholeskySolver::new()),
+            LinearSolverType::SparseQR => Box::new(linear::SparseQRSolver::new()),
+        };
+
+        // On the first iteration, we'll generate a diagonal matrix of the jacobian.
+        // Its shape will be (total_variable_dimension, total_variable_dimension).
+        // With LM, rather than solving A * dx = b for dx, we solve for (A + lambda * diag(A)) dx = b.
+        let mut jacobi_scaling_diagonal: Option<faer::sparse::SparseColMat<usize, f64>> = None;
+
+        const MIN_DIAGONAL: f64 = 1e-6;
+        const MAX_DIAGONAL: f64 = 1e32;
+
+        const INITIAL_TRUST_REGION_RADIUS: f64 = 1e4;
```

Comment:
Can you move these to `LevenbergMarquardtOptimizer`. And `impl Default for LevenbergMarquardtOptimizer` to use these default values?

Reply:
✅